In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.style as style
import seaborn as sns
import os
from turtledemo.penrose import f
style.use('ggplot')

In [5]:
print("current working directory:", os.getcwd())

current working directory: c:\Users\Kaori Komata\DATA695\creditML\notebooks\0)PreProcess


In [6]:
# Load Train.csv into DataFrame
dataPath = '../../data/credit_card_transactions.csv'
df = pd.read_csv(dataPath)
df.head(20)

Unnamed: 0 trans_date_trans_time               cc_num  \
0            0   2019-01-01 00:00:18     2703186189652095   
1            1   2019-01-01 00:00:44         630423337322   
2            2   2019-01-01 00:00:51       38859492057661   
3            3   2019-01-01 00:01:16     3534093764340240   
4            4   2019-01-01 00:03:06      375534208663984   
5            5   2019-01-01 00:04:08     4767265376804500   
6            6   2019-01-01 00:04:42       30074693890476   
7            7   2019-01-01 00:05:08     6011360759745864   
8            8   2019-01-01 00:05:18     4922710831011201   
9            9   2019-01-01 00:06:01     2720830304681674   
10          10   2019-01-01 00:06:23        4642894980163   
11          11   2019-01-01 00:06:53      377234009633447   
12          12   2019-01-01 00:06:56      180042946491150   
13          13   2019-01-01 00:07:27     5559857416065248   
14          14   2019-01-01 00:09:03     3514865930894695   
15          15   2019-01-01 00:09:20     6011999606625827   
16          16   2019-01-01 00:10:49     6011860238257910   
17          17   2019-01-01 00:10:58     3565423334076143   
18          18   2019-01-01 00:11:14     2348245054386329   
19          19   2019-01-01 00:12:34  4956828990005111019   

                                    merchant       category     amt  \
0                 fraud_Rippin, Kub and Mann       misc_net    4.97   
1            fraud_Heller, Gutmann and Zieme    grocery_pos  107.23   
2                       fraud_Lind-Buckridge  entertainment  220.11   
3         fraud_Kutch, Hermiston and Farrell  gas_transport   45.00   
4                        fraud_Keeling-Crist       misc_pos   41.96   
5           fraud_Stroman, Hudson and Erdman  gas_transport   94.63   
6                      fraud_Rowe-Vandervort    grocery_net   44.54   
7                       fraud_Corwin-Collins  gas_transport   71.65   
8                           fraud_Herzog Ltd       misc_pos    4.27   
9          fraud_Schoen, Kuphal and Nitzsche    grocery_pos  198.39   
10                    fraud_Rutherford-Mertz    grocery_pos   24.74   
11                     fraud_Kerluke-Abshire   shopping_net    7.77   
12                         fraud_Lockman Ltd    grocery_pos   71.22   
13                           fraud_Kiehn Inc    grocery_pos   96.29   
14                         fraud_Beier-Hyatt   shopping_pos    7.77   
15                    fraud_Schmidt and Sons   shopping_net    3.26   
16                    fraud_Lebsack and Sons       misc_net  327.00   
17                        fraud_Mayert Group   shopping_pos  341.67   
18  fraud_Konopelski, Schneider and Hartmann    food_dining   63.07   
19         fraud_Schultz, Simonis and Little    grocery_pos   44.71   

          first       last gender                          street  ...  \
0      Jennifer      Banks      F                  561 Perry Cove  ...   
1     Stephanie       Gill      F    43039 Riley Greens Suite 393  ...   
2        Edward    Sanchez      M        594 White Dale Suite 530  ...   
3        Jeremy      White      M     9443 Cynthia Court Apt. 038  ...   
4         Tyler     Garcia      M                408 Bradley Rest  ...   
5      Jennifer     Conner      F               4655 David Island  ...   
6        Kelsey   Richards      F     889 Sarah Station Suite 624  ...   
7        Steven   Williams      M       231 Flores Pass Suite 720  ...   
8       Heather      Chase      F     6888 Hicks Stream Suite 954  ...   
9       Melissa    Aguilar      F  21326 Taylor Squares Suite 708  ...   
10        Eddie     Mendez      M       1831 Faith View Suite 653  ...   
11      Theresa  Blackwell      F          43576 Kristina Islands  ...   
12      Charles     Robles      M                3337 Lisa Divide  ...   
13         Jack       Hill      M      5916 Susan Bridge Apt. 939  ...   
14  Christopher  Castaneda      M      1632 Cohen Drive Suite 639  ...   
15       Ronald     Carson      M             

# DATA CLEANING CHECKLIST 

## Check for missing values

In [7]:
df.isnull().sum().sort_values(ascending=False)

merch_zipcode            195973
trans_date_trans_time         0
is_fraud                      0
merch_long                    0
merch_lat                     0
unix_time                     0
trans_num                     0
dob                           0
job                           0
city_pop                      0
long                          0
lat                           0
Unnamed: 0                    0
state                         0
city                          0
street                        0
gender                        0
last                          0
first                         0
amt                           0
category                      0
merchant                      0
cc_num                        0
zip                           0
dtype: int64

Dropping after the null chek <br>
mech_zipcode shows a lot of missing values so we need to drop it <br>
Unamed: 0 is just the row index <br>

In [8]:
drop_cols = ['Unnamed: 0', 'merch_zipcode']
df.drop(columns=drop_cols, inplace=True)
print("Dropped columns:", drop_cols)

Dropped columns: ['Unnamed: 0', 'merch_zipcode']


## Data Types Check

In [9]:
print(set(df.dtypes.to_list()),"\n")
df.info()

{dtype('O'), dtype('float64'), dtype('int64')} 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1296675 entries, 0 to 1296674
Data columns (total 22 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   trans_date_trans_time  1296675 non-null  object 
 1   cc_num                 1296675 non-null  int64  
 2   merchant               1296675 non-null  object 
 3   category               1296675 non-null  object 
 4   amt                    1296675 non-null  float64
 5   first                  1296675 non-null  object 
 6   last                   1296675 non-null  object 
 7   gender                 1296675 non-null  object 
 8   street                 1296675 non-null  object 
 9   city                   1296675 non-null  object 
 10  state                  1296675 non-null  object 
 11  zip                    1296675 non-null  int64  
 12  lat                    1296675 non-null  float64
 13  long                   

## Data Description

In [10]:
df.describe()

cc_num           amt           zip           lat          long  \
count  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06   
mean   4.171920e+17  7.035104e+01  4.880067e+04  3.853762e+01 -9.022634e+01   
std    1.308806e+18  1.603160e+02  2.689322e+04  5.075808e+00  1.375908e+01   
min    6.041621e+10  1.000000e+00  1.257000e+03  2.002710e+01 -1.656723e+02   
25%    1.800429e+14  9.650000e+00  2.623700e+04  3.462050e+01 -9.679800e+01   
50%    3.521417e+15  4.752000e+01  4.817400e+04  3.935430e+01 -8.747690e+01   
75%    4.642255e+15  8.314000e+01  7.204200e+04  4.194040e+01 -8.015800e+01   
max    4.992346e+18  2.894890e+04  9.978300e+04  6.669330e+01 -6.795030e+01   

           city_pop     unix_time     merch_lat    merch_long      is_fraud  
count  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06  1.296675e+06  
mean   8.882444e+04  1.349244e+09  3.853734e+01 -9.022646e+01  5.788652e-03  
std    3.019564e+05  1.284128e+07  5.109788e+00  1.377109e+01  7.586269e-02  
min    2.300000e+01  1.325376e+09  1.902779e+01 -1.666712e+02  0.000000e+00  
25%    7.430000e+02  1.338751e+09  3.473357e+01 -9.689728e+01  0.000000e+00  
50%    2.456000e+03  1.349250e+09  3.936568e+01 -8.743839e+01  0.000000e+00  
75%    2.032800e+04  1.359385e+09  4.195716e+01 -8.023680e+01  0.000000e+00  
max    2.906700e+06  1.371817e+09  6.751027e+01 -6.695090e+01  1.000000e+00

## Duplicate Data

In [11]:
duplicates = df.duplicated().sum()
print(f"Number of duplicates: {duplicates}")

dup_count = df.duplicated().sum()
print(f"Duplicate rows: {dup_count}")
if dup_count > 0:
    df.drop_duplicates(inplace=True)
    print("Dropped duplicate rows")

Number of duplicates: 0
Duplicate rows: 0


## OUTLIER DETECTION

In [12]:
#  - Flag txns above 99th percentile
q99 = df['amt'].quantile(0.99)
#creating a new column where encoding -> where 1 indicates high value transaction outlier and 0 indicates normal transaction
df['amt_outlier'] = (df['amt'] > q99).astype(int)
#99% of all transactions are below this value which is $545.99 so only 1 % of transactions are higher than this
# might pass this encoding into a model to predict fraud (behaviour at large spending values)
print(f"99th percentile of amt = {q99:.2f}; flagged amt_outlier")

99th percentile of amt = 545.99; flagged amt_outlier


## CLASS BALANCE

In [13]:
fraud_rate = df['is_fraud'].mean()
print(f"Fraud rate: {fraud_rate:.4f} ({fraud_rate*100:.2f}%)")
# If extremely imbalanced, plan SMOTE / downsampling etc.
counts = df['is_fraud'].value_counts()
rates  = df['is_fraud'].value_counts(normalize=True)
print(pd.DataFrame({'count': counts, 'rate': rates}))

Fraud rate: 0.0058 (0.58%)
            count      rate
is_fraud                   
0         1289169  0.994211
1            7506  0.005789


# Why this matters
## Baseline “dumb” classifier

### A model that always predicts “not fraud” would be 99.42% accurate. That sounds better than it actually is since it never catches a single fraud. It would miss those cases of actual fraud causing loss. <br>


### Misleading accuracy
In imbalanced settings, accuracy is almost worthless. You need metrics that emphasize the minority (fraud) class. <br>


Choosing the right evaluation metrics <br>
Precision	Of all transactions flagged as fraud, what fraction truly were fraud? <br>
Recall	Of all actual frauds, what fraction did you catch? <br>
F1-Score	Harmonic mean of precision & recall; balances false positives & negatives. <br>
ROC-AUC	Model’s ability to rank fraud vs non-fraud across all thresholds. <br>
PR-AUC	Area under the Precision-Recall curve; more informative on imbalanced data. <br>

When training will have to split on the fraud column so that train/val/test all have the same fraud rate ~0.58%
This prevents model from seeing different fraud rates in each split. <br>

## Algorithmic Approaches <br>
### Class weights / Cost Sensitive Learning <br>
### -Tell the model that mistakes on fraud cost more than on normal transactions <br>
### -Anomaly detection which will test fraud as anomalies since train model will learn that normal transactions are not frauds (autoencoders) <br>

Note:
When fraud is rare, PR-AUC is often more meaningful than ROC-AUC, because PR-AUC focuses solely on performance for the positive (fraud) class.

# UPLOADING CLEANED DATA SET FOR PROCESSING + EDA TO USE

In [14]:
clean_path = '../../notebooks/1)Processing+EDA/credit_card_transactions_clean.csv'
df.to_csv(clean_path, index=False)
print("Cleaned data saved to", clean_path)

Cleaned data saved to ../../notebooks/1)Processing+EDA/credit_card_transactions_clean.csv
